Final/TCPServer.cpp

```diff
@@ -0,0 +1,122 @@
+    host.sin_port = htons(port); // bound to port 8080
+    host.sin_addr.s_addr = INADDR_ANY; // connected to localhost
+    const struct sockaddr *host_ptr = (struct sockaddr *) &host;
+    // bind sock_fd to port
+    if (bind(sock_fd, host_ptr, sizeof(host)) < 0)
+        error("ERROR binding to socket");
+
+    // Step 3: Listen
+    if (listen(sock_fd, BACKLOG_SIZE) < 0) // The socket will actively listen for incoming connections and queue them up
+        error("ERROR listener failed");
+
+    while (true)
+    {
+        // Reset read_fds
+        fd_set read_fds;
+        FD_ZERO(&read_fds);
+        FD_SET(sock_fd, &read_fds);
+        int maxfd = 0;
+        std::function<bool (node *)> iter_full = [&read_fds, &maxfd] (node *cur) {
+            int data = cur->data;
+            FD_SET(data, &read_fds); // populate read_fds from open_conns
+            if (data > maxfd) // calculate maxfd
+                maxfd = data;
+            return true;
+        };
+        open_conns->for_each(iter_full); // iterate through entire list
+
+        if (sock_fd > maxfd)
+            maxfd = sock_fd;
+
+        std::cout << "Attempting to Select . . ." << std::endl;
+        if (select(maxfd + 1, &read_fds, NULL, NULL, NULL) < 0) // sleeps until there is an available connection and then determines if it must be accepted or read from
+            error("My error message");
+        std::cout << "Connection Selected" << std::endl;
+
+        if (FD_ISSET(sock_fd, &read_fds))
+            accept_connection();
+        else
+            handle_connection(read_fds, handler);
+    }
+}
+
+//Pre: sock_fd is the socket bound to this server
+//Post: accept connection and add file descriptor of connected client to open_conns
+void TCPServer::accept_connection()
+{
+    // Step 4: Accecpt
+    struct sockaddr_in remote_host; // a pointer to a sockaddr struct that will be filled in with the address information of the remote host
+    socklen_t remote_len = sizeof(remote_host); // the size of the sockaddr struct in bytes
+    std::cout << "Attempting to Accept . . ." << std::endl;
+    int remotefd = accept(sock_fd, (struct sockaddr *) &remote_host, &remote_len); // dequeue from sock_fd and accept connection through remotefd
+    if (remotefd < 0)
+        error("ERROR accepting request");
+    std::cout << "Connection Accecpted" << std::endl;
+    open_conns->push(remotefd);
+}
+
+//Pre: open_conns contains fd that is also in with read_fds
+//Post: determine which fd has been updated and echo data
+void TCPServer::handle_connection(fd_set read_fds, std::function<bool (int)> handler)
+{
+    int fd;
+    std::function<bool (node *)> iter_search = [read_fds, &fd] (node *cur){
+        if (FD_ISSET(cur->data, &read_fds)) { // check if file descriptor is ready to be read from
+            fd = cur->data;
+            return false;
+        }
+        return true;
+    };
+    open_conns->for_each(iter_search);  // iterate through fdlist
+
+    if(!fd)
```

Comment:
note, it's not guaranteed that fd will be 0 if you just declare the variable `int fd`. unfortunately, for primitive types, c++ doesn't have any sort of "default constructor". so you always need to manually set the value. in addition, 0 is a valid file descriptor so i would set fd to -1 to start and check `if (fd < 0)`